lightwood/encoders/time_series/rnn.py

```diff
@@ -5,122 +5,171 @@
-                for tensor_i in range(len(data_tensor)-1):
+                next_tensor = data_tensor[:, 0, :].unsqueeze(dim=1)  # initial input
+
+                for tensor_i in range(steps - 1):
                     rand = np.random.randint(2)
-                    # teach from forward as well as from known tensor alteratively
+                    # teach from forward as well as from known tensor alternatively
                     if rand == 1:
-                        next_tensor, encoder_hidden = self._encoder.forward(data_tensor[tensor_i] , encoder_hidden)
+                        next_tensor, encoder_hidden = self._encoder.forward(
+                            data_tensor[:, tensor_i, :].unsqueeze(dim=1),
+                            encoder_hidden)
                     else:
                         next_tensor, encoder_hidden = self._encoder.forward(next_tensor.detach(), encoder_hidden)
-                    loss += criterion(next_tensor, data_tensor[tensor_i+1])
 
+                    loss += self.criterion(next_tensor, data_tensor[:, tensor_i + 1, :].unsqueeze(dim=1))
+
+                # decode
+                decoder_hidden = encoder_hidden
+                sos_vector = torch.full((batch.shape[0], 1, batch.shape[2]), self._sos,
+                                        dtype=torch.float32).to(self.device)
+                next_tensor = torch.clone(sos_vector)
+                tensor_target = torch.cat([sos_vector, batch], dim=1)
+
+                for tensor_i in range(steps - 1):
+                    rand = np.random.randint(2)
+                    # teach from forward as well as from known tensor alternatively
+                    if rand == 1:
+                        next_tensor, decoder_hidden = self._decoder.forward(
+                            tensor_target[:, tensor_i, :].unsqueeze(dim=1),
+                            decoder_hidden)
+                    else:
+                        next_tensor, decoder_hidden = self._decoder.forward(next_tensor.detach(), decoder_hidden)
+
+                    loss += self.criterion(next_tensor, tensor_target[:, tensor_i + 1, :].unsqueeze(dim=1))
 
-                loss = loss
                 average_loss += int(loss)
                 loss.backward()
-                optimizer.step()
+                self.optimizer.step()
 
-            average_loss = average_loss/len(priming_data)
+            average_loss = average_loss / len(priming_data)
 
             if average_loss < self._stop_on_error:
                 break
             if feedback_hoop_function is not None:
-                feedback_hoop_function("epoch [{epoch_n}/{total}] average_loss = {average_loss}".format(average_loss=average_loss, epoch_n=i+1, total=self._train_iters))
+                feedback_hoop_function("epoch [{epoch_n}/{total}] average_loss = {average_loss}".format(
+                    epoch_n=i + 1,
+                    total=self._train_iters,
+                    average_loss=round(average_loss, 2)))
```

Comment:
Why the rounding here ? If it's just in order to alter the stopping condition here `if average_loss < self._stop_on_error:
` then it makes sense to have it as part of the stopping condition.